File: crypto-core/ts/test/encoding.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { Writer, Reader, bytesToHex, hexToBytes, utf8 } from "../src/bytes.js";
import { encodeRecord, decodeRecord, ImageKind, encodeMessage, Op } from "../src/message.js";

test("Writer/Reader round-trips primitives and length-prefixed fields", () => {
  const buf = new Writer().u8(1).u16(258).u32(0xdeadbeef).u64(0x0102030405060708n).bytes(utf8("hi")).str("héllo").finish();
  const rd = new Reader(buf);
  assert.equal(rd.u8(), 1);
  assert.equal(rd.u16(), 258);
  assert.equal(rd.u32(), 0xdeadbeef);
  assert.equal(rd.u64(), 0x0102030405060708n);
  assert.equal(bytesToHex(rd.bytes()), bytesToHex(utf8("hi")));
  assert.equal(rd.str(), "héllo");
  rd.end();
});

test("Reader rejects truncation, trailing bytes, bad utf8, oversized length", () => {
  assert.throws(() => new Reader(new Uint8Array([0x00])).u32(), /truncated/);
  const w = new Writer().u8(7).finish();
  const rd = new Reader(new Uint8Array([...w, 0xff]));
  assert.equal(rd.u8(), 7);
  assert.throws(() => rd.end(), /trailing/);
  // length prefix says 10 bytes but buffer has 1
  assert.throws(() => new Reader(new Uint8Array([0, 0, 0, 10, 1])).bytes(), /truncated/);
  // invalid utf8 (0xff is not valid)
  assert.throws(() => new Reader(new Uint8Array([0, 0, 0, 1, 0xff])).str(), /./);
});

test("record encode/decode round-trip", () => {
  const rec = {
    streamId: utf8("ledger.invoices"),
    message: { tableId: "ledger.invoices", rowId: hexToBytes("0001"), columnId: "amount", op: Op.UPDATE, seq: 7n, prevTxid: hexToBytes("ab".repeat(32)) },
    imageKind: ImageKind.COMMITMENT,
    changeImage: hexToBytes("cd".repeat(32)),
    tag: hexToBytes("ef".repeat(32)),
  };
  const enc = encodeRecord(rec);
  const dec = decodeRecord(enc);
  assert.equal(bytesToHex(encodeRecord(dec)), bytesToHex(enc));
  assert.equal(dec.message.seq, 7n);
  assert.equal(dec.message.columnId, "amount");
  assert.equal(dec.imageKind, ImageKind.COMMITMENT);
});

test("record decode rejects bad magic and trailing bytes", () => {
  const enc = encodeRecord({
    streamId: utf8("s"),
    message: { tableId: "t", rowId: new Uint8Array(0), columnId: "c", op: Op.INSERT, seq: 0n, prevTxid: new Uint8Array(0) },
    imageKind: ImageKind.PLAINTEXT,
    changeImage: utf8("v"),
    tag: hexToBytes("00".repeat(32)),
  });
  const bad = enc.slice();
  bad[0] = 0x58;
  assert.throws(() => decodeRecord(bad), /magic/);
  assert.throws(() => decodeRecord(new Uint8Array([...enc, 0x00])), /trailing/);
});

test("encodeMessage rejects malformed prevTxid", () => {
  assert.throws(
    () => encodeMessage({ tableId: "t", rowId: new Uint8Array(0), columnId: "c", op: Op.INSERT, seq: 0n, prevTxid: new Uint8Array(5) }),
    /prevTxid/,
  );
});

File: crypto-core/go/encoding_test.go
package cryptocore

import (
	"bytes"
	"testing"
)

func TestWriterReaderRoundTrip(t *testing.T) {
	buf := NewWriter().U8(1).U16(258).U32(0xdeadbeef).U64(0x0102030405060708).Bytes([]byte("hi")).Str("héllo").Finish()
	rd := NewReader(buf)
	if rd.U8() != 1 || rd.U16() != 258 || rd.U32() != 0xdeadbeef || rd.U64() != 0x0102030405060708 {
		t.Fatal("primitive mismatch")
	}
	if !bytes.Equal(rd.Bytes(), []byte("hi")) {
		t.Fatal("bytes mismatch")
	}
	if rd.Str() != "héllo" {
		t.Fatal("str mismatch")
	}
	if err := rd.End(); err != nil {
		t.Fatalf("end: %v", err)
	}
}

func TestReaderRejections(t *testing.T) {
	if rd := NewReader([]byte{0x00}); func() bool { rd.U32(); return rd.Err() == nil }() {
		t.Fatal("expected truncation error")
	}
	// trailing bytes
	rd := NewReader([]byte{7, 0xff})
	rd.U8()
	if err := rd.End(); err == nil {
		t.Fatal("expected trailing-bytes error")
	}
	// length prefix overruns buffer
	rd2 := NewReader([]byte{0, 0, 0, 10, 1})
	rd2.Bytes()
	if rd2.Err() == nil {
		t.Fatal("expected truncation on oversized length")
	}
	// invalid utf-8
	rd3 := NewReader([]byte{0, 0, 0, 1, 0xff})
	rd3.Str()
	if rd3.Err() == nil {
		t.Fatal("expected invalid utf-8 error")
	}
}

func TestRecordRoundTripAndRejections(t *testing.T) {
	rec := FieldRecord{
		StreamID:    []byte("ledger.invoices"),
		Seq:         7,
		PrevTxid:    bytes.Repeat([]byte{0xab}, 32),
		ImageKind:   ImageCommitment,
		ChangeImage: bytes.Repeat([]byte{0xcd}, 32),
		Tag:         bytes.Repeat([]byte{0xef}, 32),
	}
	enc, err := EncodeRecord(rec)
	if err != nil {
		t.Fatal(err)
	}
	dec, err := DecodeRecord(enc)
	if err != nil {
		t.Fatal(err)
	}
	if dec.Seq != 7 || dec.ImageKind != ImageCommitment {
		t.Fatal("decoded fields mismatch")
	}
	reenc, _ := EncodeRecord(dec)
	if !bytes.Equal(reenc, enc) {
		t.Fatal("re-encode mismatch")
	}
	// bad magic
	bad := append([]byte{}, enc...)
	bad[0] = 0x58
	if _, err := DecodeRecord(bad); err == nil {
		t.Fatal("expected bad-magic error")
	}
	// trailing bytes
	if _, err := DecodeRecord(append(append([]byte{}, enc...), 0x00)); err == nil {
		t.Fatal("expected trailing-bytes error")
	}
}

func TestEncodeMessageRejectsBadPrevTxid(t *testing.T) {
	_, err := EncodeMessage(ChangeMessage{TableID: "t", ColumnID: "c", Op: OpInsert, Seq: 0, PrevTxid: make([]byte, 5)})
	if err == nil {
		t.Fatal("expected prevTxid length error")
	}
}
